hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java

```diff
@@ -520,49 +535,65 @@ private List<BalanceChange> constructBalanceChanges(final CryptoTransferWrapper
                                     BalanceChange.changingFtUnits(
                                             Id.fromGrpcToken(fungibleTransfer.getDenomination()),
                                             fungibleTransfer.getDenomination(),
                                             aaWith(
                                                     fungibleTransfer.sender(),
                                                     -fungibleTransfer.amount(),
                                                     fungibleTransfer.isApproval()),
-                                            EntityIdUtils.accountIdFromEvmAddress(senderAddress))));
+                                            accountId)));
                 } else if (fungibleTransfer.sender() == null) {
-                    changes.add(
+                    allChanges.add(
                             BalanceChange.changingFtUnits(
                                     Id.fromGrpcToken(fungibleTransfer.getDenomination()),
                                     fungibleTransfer.getDenomination(),
                                     aaWith(
                                             fungibleTransfer.receiver(),
                                             fungibleTransfer.amount(),
                                             fungibleTransfer.isApproval()),
-                                    EntityIdUtils.accountIdFromEvmAddress(senderAddress)));
+                                    accountId));
                 } else {
-                    changes.add(
+                    allChanges.add(
                             BalanceChange.changingFtUnits(
                                     Id.fromGrpcToken(fungibleTransfer.getDenomination()),
                                     fungibleTransfer.getDenomination(),
                                     aaWith(
                                             fungibleTransfer.sender(),
                                             -fungibleTransfer.amount(),
                                             fungibleTransfer.isApproval()),
-                                    EntityIdUtils.accountIdFromEvmAddress(senderAddress)));
+                                    accountId));
                 }
             }
-            if (changes.isEmpty()) {
-                for (final var nftExchange : tokenTransferWrapper.nftExchanges()) {
-                    changes.add(
-                            BalanceChange.changingNftOwnership(
-                                    Id.fromGrpcToken(nftExchange.getTokenType()),
-                                    nftExchange.getTokenType(),
-                                    nftExchange.asGrpc(),
-                                    EntityIdUtils.accountIdFromEvmAddress(senderAddress)));
-                }
+            for (final var nftExchange : tokenTransferWrapper.nftExchanges()) {
+                allChanges.add(
+                        BalanceChange.changingNftOwnership(
```

Comment:
Q: Should there be an isApproval flag for NTS exchanges here?

Reply:
The isApproval flag is set inside changingNftOwership [method] https://github.com/hashgraph/hedera-services/blob/8776416c86cdf1f1450ec5fc6cbb961e1f78ddbe/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/ledger/BalanceChange.java#L109